bindings/rust/s2n-tls/src/raw/pool.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Utilities to handle reusing connections.
+//!
+//! Creating a single new connection requires significant
+//! memory allocations (about 50-60 KB, according to some tests).
+//! Instead of allocating memory for a new connection, existing
+//! memory can be reused by calling
+//! [Connection::wipe()](`s2n_tls::raw::connection::Connection::wipe()).
+//!
+//! The [`Pool`] trait allows applications to define an
+//! [Object pool](https://en.wikipedia.org/wiki/Object_pool_pattern) that
+//! wipes and stores connections after they are dropped.
+//!
+//! We also provide a basic Pool implementation, [`ConfigPool`], that
+//! implements the pool as a [VecDeque](`std::collections::VecDeque`)
+//! with a fixed maximum size.
+
+use crate::raw::{
+    config::Config,
+    connection::{AsConnection, Connection},
+    enums::Mode,
+    error::Error,
+};
+use std::{
+    collections::VecDeque,
+    sync::{Arc, Mutex},
+};
+
+/// A connection produced by a [`Pool`].
+///
+/// When dropped, returns ownership of the connection to
+/// the pool that produced it by calling [`Pool::reclaim()`].
+#[derive(Debug)]
+pub struct PooledConnection<T: Pool> {
+    pool: T,
```

Comment:
I don't love requiring dynamic dispatch :\

It might be too weird, but how about [this](https://github.com/aws/s2n-tls/pull/3336/commits/4b23ecb436a5f822841d3279049d6877aa20fe99)? If the application wants to use `Arc<dyn Pool>`, then they don't need to thread the `T` everywhere, but not using dynamic dispatch is still an option.